### PR TITLE
[RFC] bwrap-oci: wrapper to run bwrap from an OCI configuration file

### PR DESCRIPTION
**Pull request by giuseppe**
best effort conversion from OCI specs to bubblewrap command line.

It will be useful for supporting: https://github.com/projectatomic/atomic/pull/483

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

### CODE REVIEW

**Comment by rhatdan** [comment]
I would have written this in python.  Easier json parsing, and easier to contribute.  But nice effort.

I think something like this needs a bunch of tests, different OCI files and expected generated command lines. 

@mrunalp @runcom PTAL

**Comment by runcom** [comment]
> I would have written this in python. Easier json parsing, and easier to contribute. But nice effort.

right, any particular reason why this is written directly in C?

File: bwrap-oci.c [hunk 1]
```diff
@@ -0,0 +1,500 @@
...
+  g_print ("\n");
+}
+
+static void
+generate_seccomp_rules_file (struct context *context)
+{
+  if (context->seccomp)
+    {
+      char fdstr[10];
+      int fd = open (".", O_TMPFILE | O_RDWR, S_IRUSR | S_IWUSR);
```
Comment (runcom) [inline bwrap-oci.c]: should you close this file at the end?

File: bwrap-oci.c [hunk 2]
```diff
@@ -0,0 +1,500 @@
...
+          g_free (val);
+          g_variant_unref (env);
+        }
+    }
+  if (json_object_has_member (root, "selinuxLabel"))
+    {
+      JsonNode *label = json_object_get_member (root, "selinuxLabel");
+      collect_options (con, "--exec-label", json_node_get_string (label), NULL);
+    }
+  if (json_object_has_member (root, "user"))
```
Comment (rhatdan) [inline bwrap-oci.c]: Needs processing of fileLabel.

```
 if (json_object_has_member (root, "mountLabel"))
 +    {
 +      JsonNode *label = json_object_get_member (root, "mountLabel");
 +      collect_options (con, "--file-label", json_node_get_string (label), NULL);
 +    }
```

**Comment by rhatdan** [comment]
Doing some quick grepping. These are the options in brwap that are not in your parser.  Not sure how many matter.

```
--args
--bind-data
--dev
--dir
--file
--file-label
--lock-file
--ro-bind
--share-user
--symlink
--sync-fd
--tmpfs
--unsetenv
```

**Comment by giuseppe** [comment]
I've started writing it in Python, until I've realized libseccomp under Fedora/RHEL has not the Python bindings and that they require cython, then I switched to a C implementation.

**Comment by runcom** [comment]
@giuseppe for what are you in need of libseccomp here?

File: bwrap-oci.c (hunk 1 above)
Comment (giuseppe) [inline bwrap-oci.c]: the temporary file will be used by bubblewrap to read the seccomp rules:

`bwrap ... --seccomp FD ...`

so we need to keep it open in order to be used by bubblewrap

**Comment by runcom** [comment]
@giuseppe seccomp in golang is well done and to convert from OCI spec to raw seccomp I believe you can do it easily in go

File: bwrap-oci.c (hunk 1 above)
Comment (runcom) [inline bwrap-oci.c]: i see

File: bwrap-oci.c (hunk 2 above)
Comment (giuseppe) [inline bwrap-oci.c]: thanks, I have pushed a fixup patch to handle it

**Comment by cgwalters** [comment]
I'd say golang is probably going to be the best bet since it'll make it easier to reuse code.  The major open question is whether this should live in the bubblewrap project or somewhere else.  It seems like it's easiest for this to live in the atomic command which already has some go (albeit a trivial amount)?

**Comment by mrunalp** [comment]
I agree that go would be a good choice given that the ecosystem tooling is all in go.

**Comment by runcom** [comment]
Can't be just it another project under projectatomic? Easily reusable (as we did with skopeo, image, storage and the like)

**Comment by cgwalters** [comment]
"bubblewrap-oci"?

**Comment by runcom** [comment]
I suck at naming things but yeah I like that ha

**Comment by giuseppe** [comment]
@cgwalters I thought this was the right place for the new tool as it can track bubblewrap development more easily and we won't have to worry what options are supported or not.  We can achieve that in another way, but this was the easiest and quickest way

**Comment by cgwalters** [comment]
@giuseppe I'm a big fan of the effort!  We definitely want to more align bubblewrap/OCI.  I think we can pretty easily version lock bubblewrap and bubblewrap-oci even if they're stored in separate repos; use RPM dependencies etc. right?

**Comment by giuseppe** [comment]
@cgwalters sure :)  Can we create a new project?  Is it fine to keep the C implementation for now so that we can start using for user containers?  It should not make any difference for users if we decide to rewrite it in golang later.

**Comment by rhatdan** [comment]
I have no problem aligning it with bubblewrap or having it be separate.  I would see this shipping with bubblewrap rpm, not sure why having a separate RPM would be helpful.  Don't figure it will have much size (until it is rewritten in go :^)

We can ship it as 'C' for now, but rewriting this in go would be good for the cool kids...  And probably be easier to contribute to and simpler to maintain.

**Comment by cgwalters** [comment]
A separate package mostly avoids mixing build systems in one git repo which simplies things.  Also at the moment, rpm-ostree only depends on C code (including bubblewrap), and I have a preference to keep it that way.

**Comment by giuseppe** [comment]
I created a separated project temporarily at: https://github.com/giuseppe/bwrap-oci

**Comment by giuseppe** [comment]
since bwrap-oci is going to be a project on its own, I am closing this PR

**Pull request closed without merging**